### PR TITLE
Add preflight github action

### PR DESCRIPTION
**Pull request by sethrj**
This adds a preflight build for testing Celeritas cached on native GitHub hardware before proceeding to the full build. Because it uses ccache, it should only take a few seconds for each push, versus the 12 minutes or so for the cuda/hip builds (at least 2 minutes of which is downloading the docker images).

It uses a release build (to avoid debug symbols that increase cache size) and debug assertions (to catch simple issues during testing).

### CODE REVIEW

File: .github/workflows/build-local.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Build directly on the GitHub runner
+name: build-local
+on:
+  workflow_dispatch:
+  workflow_call:
+
+concurrency:
+  group: build-${{github.ref}}-${{github.event.pull_request.number || github.run_number}}-${{github.workflow}}
```
Comment (pcanal) [inline .github/workflows/build-local.yml]: For the group, I would replace `build-` with `build-local-` or `prebuild-`, to differentiate it from the full build.

File: .github/workflows/build-local.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          sudo apt-get -y update
+          sudo apt-get -y install \
+            ccache cmake ninja-build python3 \
+            nlohmann-json3-dev libgtest-dev libhepmc3-dev \
+            ${{matrix.compiler}}
+      - name: Check out Celeritas
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 255
+          fetch-tags: true
```
Comment (pcanal) [inline .github/workflows/build-local.yml]: Do we need the tag and the 255 depth for the quick build?  Wouldn't the already downloaded by github repo be sufficient?

File: .github/workflows/build-local.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            -DCMAKE_CXX_FLAGS="-Wall -Wextra -pedantic -Werror -Wno-error=deprecated-declarations" \
+            ..
+      - name: Build all
+        working-directory: build
+        run: |
+          ninja
+      - name: Run tests
+        working-directory: build
+        run: |
+          ctest --parallel 2 --timeout 15 --output-on-failure \
```
Comment (pcanal) [inline .github/workflows/build-local.yml:84]: Is the number of core given by github known?

**Review by pcanal — CHANGES_REQUESTED** [review]
group name needs review.  Other comments are potential optimization.

File: .github/workflows/build-local.yml (hunk 2 above)
Comment (sethrj) [reply .github/workflows/build-local.yml]: Actually good point, we don't . It was only truly necessary for the "install" step of the full thing.

File: .github/workflows/build-local.yml (hunk 3 above)
Comment (sethrj) [reply .github/workflows/build-local.yml:84]: I don't think so: I looked it up, and it just had a hard-coded info page about the github runners.

Comment (pcanal) [reply .github/workflows/build-local.yml:84]: It is probably not enough of a gain to worry about but we could also determine the number of core at run-time (https://github.com/orgs/community/discussions/25057)

**Review by pcanal — APPROVED** [review]
Looks good (albeit maybe a future improve to increase the parallelism of ctest).  ccache is working as hope for (first build 5 minutes, second identical build a few seconds).

File: .github/workflows/build-local.yml (hunk 3 above)
Comment (sethrj) [reply .github/workflows/build-local.yml:84]: Actually I forgot I added a script earlier (scripts/ci/host-system-info.cmake) that's meant to pull this number in. We'll just have to enable it.

**Pull request merged**